subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/JavaToolchainRepository.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain;
+
+import org.gradle.api.Incubating;
+
+import java.net.URI;
+import java.util.Optional;
+
+/**
+ * Java Toolchain Download SPI.
+ *
+ * @since 7.6
+ */
+@Incubating
+public interface JavaToolchainRepository {
+
+    /**
+     * Returns the URI from which a Java Toolchain matching the provided specification can be downloaded.
+     * The URI must point to either a ZIP or a TAR archive file.
+     *
+     * Returns an empty Optional if and only if the provided specification can't be matched.
+     */
+    Optional<URI> toUri(JavaToolchainSpec spec);
+
+    /**
+     * Returns information about the Java Toolchain that would match the provided specification.
+     *
+     * Returns an empty Optional if and only if the provided specification can't be matched.
+     */
+    Optional<Metadata> toMetadata(JavaToolchainSpec spec);
```

Comment:
It might be better to have a single entry point that returns an `Optional<SomeObject>` and `SomeObject` carries the `URI` and `Metadata` information